### PR TITLE
feat: Implement AUTHENTICATE flow

### PR DESCRIPTION
**Pull request by duesee**
Closes #71 

I already merged all minor changes, typos, reformatting, etc. on main. This branch is a cleaned-up (and rebased) version of `git diff duesee_issues_16_v2`.

Multiple things happen here (see commits):

* TaskMap (could be a separate PR)
* AUTHENTICATE for imap-flow
  * Please look at the fix in particular. I think we accidentally introduced a bug during the "Dummy refactoring".
* AUTHENTICATE for proxy (could be a separate PR)
* AUTHENTICATE for tasks (could be a separate PR)

### CODE REVIEW

File: examples/server_authenticate.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    };
+
+    let mut current_authenticate_tag = None;
+
+    loop {
+        let event = server.progress().await?;
+        println!("{event:?}");
+
+        // We don't implement any real SASL mechanism in this example.
+        let pretend_to_need_more_data = rand::random();
```
Comment (jakoschiko) [inline examples/server_authenticate.rs:35]: I don't understand this. If `pretend_to_need_more_data` is true, then `authenticate_finish` is never called. Is this a useful example?

File: src/server.rs [hunk 2]
```diff
@@ -271,9 +394,16 @@ pub enum ServerFlowEvent {
         /// Formerly enqueued [`Response`] that was now sent.
         response: Response<'static>,
     },
-    CommandReceived {
-        command: Command<'static>,
+    /// Command received.
+    CommandReceived { command: Command<'static> },
+    /// Command AUTHENTICATE received.
```
Comment (jakoschiko) [inline src/server.rs:399]: This doc should mention that either `authenticate_continue` or `authenticate_finish` should be called afterwards.

File: src/server.rs [hunk 3]
```diff
@@ -271,9 +394,16 @@ pub enum ServerFlowEvent {
...
+    CommandAuthenticateReceived {
+        command_authenticate: CommandAuthenticate,
     },
+    /// Continuation to AUTHENTICATE received.
+    ///
+    /// Note: This can either mean `Continue` or `Cancel` depending on `authenticate_data`.
+    AuthenticateDataReceived { authenticate_data: AuthenticateData },
```
Comment (jakoschiko) [inline src/server.rs:406]: This doc should mention that either `authenticate_continue` or `authenticate_finish` should be called afterwards.

File: src/client.rs [hunk 4]
```diff
@@ -267,17 +378,36 @@ pub enum ClientFlowEvent {
         /// (e.g. [`Code::Alert`](imap_codec::imap_types::response::Code::Alert)).
         status: Status<'static>,
     },
+    AuthenticateStarted {
+        handle: ClientFlowCommandHandle,
+    },
+    /// Server is requesting (more) authentication data.
```
Comment (jakoschiko) [inline src/client.rs:384]: The doc should mention that `authenticate_continue` should be called afterwards.

File: examples/server_authenticate.rs (hunk 1 above)
Comment (duesee) [reply examples/server_authenticate.rs:35]: `pretend_to_need_more_data` is randomly assigned in the loop. What should happen is that you get a variety of different SASL flows randomly. If `pretend_to_need_more_data` is true, we send a continuation and ask for more. Then, we proceed due to the client sending `AuthenticateData` below and eventually finish when `pretend_to_need_more_data` is false,

**Review by jakoschiko — APPROVED** [review]

**Pull request merged**